test/common/config/delta_subscription_state_test.cc

```diff
@@ -0,0 +1,347 @@
+      const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+      const std::string& version_info, absl::optional<std::string> nonce = absl::nullopt) {
+    envoy::api::v2::DeltaDiscoveryResponse message;
+    *message.mutable_resources() = added_resources;
+    *message.mutable_removed_resources() = removed_resources;
+    message.set_system_version_info(version_info);
+    if (nonce.has_value()) {
+      message.set_nonce(nonce.value());
+    }
+    EXPECT_CALL(callbacks_, onConfigUpdate(_, _, _)).Times(1);
+    return state_.handleResponse(message);
+  }
+
+  UpdateAck deliverBadDiscoveryResponse(
+      const Protobuf::RepeatedPtrField<envoy::api::v2::Resource>& added_resources,
+      const Protobuf::RepeatedPtrField<std::string>& removed_resources,
+      const std::string& version_info, std::string nonce) {
+    envoy::api::v2::DeltaDiscoveryResponse message;
+    *message.mutable_resources() = added_resources;
+    *message.mutable_removed_resources() = removed_resources;
+    message.set_system_version_info(version_info);
+    message.set_nonce(nonce);
+    EXPECT_CALL(callbacks_, onConfigUpdate(_, _, _)).WillOnce(Throw(EnvoyException("oh no")));
+    return state_.handleResponse(message);
+  }
+
+  NiceMock<MockSubscriptionCallbacks<envoy::api::v2::Cluster>> callbacks_;
+  NiceMock<LocalInfo::MockLocalInfo> local_info_;
+  NiceMock<Event::MockDispatcher> dispatcher_;
+  Stats::IsolatedStoreImpl store_;
+  SubscriptionStats stats_;
+  // We start out interested in three resources: name1, name2, and name3.
+  DeltaSubscriptionState state_;
+};
+
+Protobuf::RepeatedPtrField<envoy::api::v2::Resource>
+populateRepeatedResource(std::vector<std::pair<std::string, std::string>> items) {
+  Protobuf::RepeatedPtrField<envoy::api::v2::Resource> add_to;
+  for (const auto& item : items) {
+    auto* resource = add_to.Add();
+    resource->set_name(item.first);
+    resource->set_version(item.second);
+  }
+  return add_to;
+}
+
+// Basic gaining/losing interest in resources should lead to (un)subscriptions.
+TEST_F(DeltaSubscriptionStateTest, SubscribeAndUnsubscribe) {
+  {
+    state_.updateResourceInterest({"name2", "name3", "name4"}); // drop name1, add name4
+    envoy::api::v2::DeltaDiscoveryRequest cur_request = state_.getNextRequest();
+    EXPECT_THAT(cur_request.resource_names_subscribe(), UnorderedElementsAre("name4"));
+    EXPECT_THAT(cur_request.resource_names_unsubscribe(), UnorderedElementsAre("name1"));
+  }
+  {
+    state_.updateResourceInterest({"name1", "name2"}); // add back name1, drop name3 and 4
+    envoy::api::v2::DeltaDiscoveryRequest cur_request = state_.getNextRequest();
+    EXPECT_THAT(cur_request.resource_names_subscribe(), UnorderedElementsAre("name1"));
+    EXPECT_THAT(cur_request.resource_names_unsubscribe(), UnorderedElementsAre("name3", "name4"));
+  }
+}
+
+// Delta xDS reliably queues up and sends all discovery requests, even in situations where it isn't
```

Comment:
These comments are super awesome 💯

Reply:
Thanks! It occurred to me that while I pointed this "resubscribe" thing out in the protocol spec doc, I didn't in discovery.proto, so I added a comment there, too.